python/tank/authentication/console_authentication.py

```diff
@@ -68,47 +79,149 @@ def authenticate(self, hostname, login, http_proxy):
+            # the code is invalid or already used, it will be caught by the except clause beneath.
+            return (
+                hostname,
+                login,
+                session_cache.generate_session_token(
+                    hostname, login, password, http_proxy, auth_token=code
+                ),
+                None,
+            )
+
+    def _authenticate_unified_login_flow2(self, hostname, login, http_proxy):
+        print()
+        print(
+            "The authentication to {sg_url} requires to use your web browser.\n"
+            "\n"
+            'After pressing "continue", your web browser will open shortly '
+            "targeting your selected ShotGrid site.\n"
+            "\n"
+            "If you are not already authenticated to {sg_url} in the browser,"
+            "you will first need to authenticate.\n"
+            "\n"
+            "Then, you will be prompted to review the access request.\n"
+            'Select "Approve" and come back to this application.'
+            "\n".format(sg_url=hostname)
+        )
+
+        self._read_clean_input("Press enter when you are ready to continue")
+        print("\n")  # Always have 2 empty lines after a prompt
+        print(
+            "Your browser will open shortly.\n"
+            "Once you approved the access request, come back to this "
+            "application"
+        )
+        print()
+        session_info = ulf2_authentication.process(
+            hostname,
+            http_proxy=http_proxy,
+            product="toolkit",  # Same as "PRODUCT_IDENTIFIER" from LoginDialog
+            browser_open_callback=lambda u: webbrowser.open(u),
+        )
+
+        print()
+        if not session_info:
+            raise AuthenticationError("The web authentication failed.")
+
+        print("The web authentication succeed, now processing.")
+        return session_info
+
+    def _get_auth_method(self, hostname, http_proxy):
+        if not is_unified_login_flow2_enabled_on_site(hostname, http_proxy):
+            return self._authenticate_legacy
+
+        if is_autodesk_identity_enabled_on_site(
+            hostname, http_proxy
+        ) or is_sso_enabled_on_site(hostname, http_proxy):
+            return self._authenticate_unified_login_flow2
+
+        # We have 2 choices here
+        # Let's see which method the user chose previously for this site
+        # Then prompt them to chose
+
+        method_saved = session_cache.get_preferred_method(hostname)
+        print(
+            "\n"
+            "The ShotGrid site support two authentication methods:\n"
+            " 1. App Session Launcher using your local web browser\n"
+            " 2. Legacy method using login/password\n"
+        )
+
+        method = self._get_keyboard_input(
+            "Select a method (1 or 2)",
+            default_value="1" if method_saved == "credentials" else "2",
+        )
```

Comment:
Maybe we can define a dictionary to avoid having magic values. Something like:

```py
KEYBOARD_METHOD_MAP = {
    "credentials": "1",
    "unified_login_flow2": "2",
}
```

So this can become 

```suggestion
        method = self._get_keyboard_input(
            "Select a method (1 or 2)",
            default_value=KEYBOARD_METHOD_MAP.get(method_saved, "2"),
        )
```

I'd love to use Enums but we can affect legacy code.